### PR TITLE
refactor xml

### PR DESCRIPTION
**Pull request by andiwand**

### CODE REVIEW

**Comment by andiwand** [comment]
@TomTasche small update - this branch is passing our checks. next step is to compare the output.

File: odf/src/Meta.cpp [hunk 1]
```diff
@@ -99,40 +100,37 @@ bool lookupStartKeyTypes(const std::string &checksum,
...
-                             const std::uint32_t limitRows,
+void estimateTableDimensions(const pugi::xml_node &table, std::uint32_t &rows,
+                             std::uint32_t &cols, const std::uint32_t limitRows,
                              const std::uint32_t limitCols) {
   rows = 0;
   cols = 0;
 
   common::TableCursor tl;
 
   // TODO we dont need to recurse so deep
```
Comment (andiwand) [inline odf/src/Meta.cpp:111]:
```suggestion
```

File: odf/src/OpenDocument.cpp [hunk 2]
```diff
@@ -109,25 +88,23 @@ void generateContent_(tinyxml2::XMLHandle &in, Context &context) {
...
-                context.config->entryOffset + context.config->entryCount))) {
-            ContentTranslator::html(c, context);
-          } else {
-            ++context.entry; // TODO hacky
-          }
-          ++i;
-        });
+    for (auto &&e : content) {
+      if (e.name() != entryName)
+        return;
```
Comment (andiwand) [inline odf/src/OpenDocument.cpp:96]:
```suggestion
        continue;
```

File: ooxml/src/PresentationTranslator.cpp [hunk 3]
```diff
@@ -4,97 +4,95 @@
...
 
-void MarginAttributesTranslator(const tinyxml2::XMLElement &in,
-                                std::ostream &out, Context &) {
-  const tinyxml2::XMLAttribute *marLAttr = in.FindAttribute("marL");
-  if (marLAttr != nullptr) {
-    float marLIn = marLAttr->Int64Value() / 914400.0f;
+void MarginAttributesTranslator(const pugi::xml_node &in, std::ostream &out,
+                                Context &) {
+  const auto marLAttr = in.attribute("marL");
+  if (!marLAttr) {
```
Comment (andiwand) [inline ooxml/src/PresentationTranslator.cpp:78]:
```suggestion
  if (marLAttr) {
```

File: ooxml/src/PresentationTranslator.cpp [hunk 4]
```diff
@@ -4,97 +4,95 @@
...
+    float marLIn = marLAttr.as_float() / 914400.0f;
     out << "margin-left:" << marLIn << "in;";
   }
 
-  const tinyxml2::XMLAttribute *marRAttr = in.FindAttribute("marR");
-  if (marRAttr != nullptr) {
-    float marRIn = marRAttr->Int64Value() / 914400.0f;
+  const auto marRAttr = in.attribute("marR");
+  if (!marRAttr) {
```
Comment (andiwand) [inline ooxml/src/PresentationTranslator.cpp:84]:
```suggestion
  if (marRAttr) {
```

File: ooxml/src/PresentationTranslator.cpp [hunk 5]
```diff
@@ -238,18 +225,16 @@ void ParagraphTranslator(const tinyxml2::XMLElement &in, std::ostream &out,
...
-              if (common::StringUtil::endsWith(e1.Name(), "Pr"))
-                ;
-              else if (std::strcmp(e1.Name(), "w:r") != 0)
-                empty = false;
-              else if (common::StringUtil::endsWith(e2.Name(), "Pr"))
-                empty = false;
-            });
-      });
+  for (auto &&e1 : in) {
+    for (auto &&e2 : in) {
```
Comment (andiwand) [inline ooxml/src/PresentationTranslator.cpp:229]:
```suggestion
    for (auto &&e2 : e1) {
```

File: ooxml/src/PresentationTranslator.cpp [hunk 6]
```diff
@@ -301,23 +282,19 @@ void TableTranslator(const tinyxml2::XMLElement &in, std::ostream &out,
...
   ElementAttributeTranslator(in, out, context);
 
-  const tinyxml2::XMLElement *ref =
-      tinyxml2::XMLHandle((tinyxml2::XMLElement &)in)
-          .FirstChildElement("p:blipFill")
-          .FirstChildElement("a:blip")
-          .ToElement();
-  if (ref == nullptr || ref->FindAttribute("r:embed") == nullptr) {
+  const auto ref = in.child("p:blipFill").child("a:blip");
+  if (ref || ref.attribute("r:embed")) {
```
Comment (andiwand) [inline ooxml/src/PresentationTranslator.cpp:291]:
```suggestion
  if (!ref || !ref.attribute("r:embed")) {
```

**Comment by andiwand** [comment]
output compared
- [ ] meta
  - [x] odt
  - [ ] odp
  - [ ] ods
  - [ ] odg
  - [ ] docx
  - [ ] pptx
  - [ ] xlsx
- [ ] html
  - [x] odt
  - [x] odp
  - [ ] ods
  - [x] odg
  - [x] docx
  - [x] pptx
  - [ ] xlsx

**Comment by andiwand** [comment]
ready :tada:

File: odf/src/ContentTranslator.cpp [hunk 7]
```diff
@@ -64,86 +64,81 @@ void StyleClassTranslator(const tinyxml2::XMLElement &in, std::ostream &out,
...
     const auto it = context.defaultCellStyles.find(context.tableCursor.col());
     if (it != context.defaultCellStyles.end()) {
       StyleClassTranslator(it->second, out, context);
       out << " ";
     }
   }
-  const auto valueTypeAttr = in.FindAttribute("office:value-type");
-  if (valueTypeAttr != nullptr) {
-    out << "odr-value-type-" << valueTypeAttr->Value() << " ";
+  if (const auto valueTypeAttr = in.attribute("office:value-type");
```
Comment (TomTasche) [inline odf/src/ContentTranslator.cpp:74]: what is this crazy syntax?

Comment (andiwand) [reply odf/src/ContentTranslator.cpp:74]: c++17 😄 i think it is quite cool. similar to `for`-loop syntax

**Pull request merged**